library-best-practices/JLBP-1.md

```diff
@@ -0,0 +1,20 @@
+[JLBP-1] Minimize dependencies
+------------------------------
+
+Use the minimum number of dependencies that is reasonable.
+Adding a dependency for a large amount of functionality may be ok,
+but avoid pulling in dependencies just to save a few lines of code,
+because every dependency is a liability.
+
+Some specific notes about minimizing dependencies:
+
+- Use the smallest scope possible. For example, auto-value doesn't
+  need to use compile scope, and can instead use compile-only,
+  since it doesn't need to appear on the classpath of consumers.
```

Comment:
I wouldn't recommend `compile-only` for `auto-value` and other annotations artifacts. It's caused problems for Guava users: See https://github.com/google/guava/issues/2721 and various other bugs files against us.

For AutoValue specifically, I would recommend [the pattern in its user guide](https://github.com/google/auto/blob/master/value/userguide/index.md#in-pomxml), in which `auto-value-annotations` is a `compile` dependency and `auto-value` itself is an annotation processor (which apparently you can accomplish by making it `provided`, but that seems like not the most precise way).

I can send a PR if you'd like. Maybe you _want_ it to include an example of when `compile-only` is the right thing? If so, maybe that's the servlet API case?

Reply:
I'm not familiar with auto-value details, but servlets are definitely not compile-only.

Reply:
Ah, I am going off Ray's comment in an internal thread. It sounds like I can expect to hear a response from you there later :)

Reply:
@cpovirk a PR would be much appreciated! If we don't truly have any examples where we should use compile-only, then we should probably remove the mention of it.